Return opaque origin for blob: URL containing inner non-http(s): URL

annevk opened the pull request:
#### b1a659b38f09d05fd3e7b3b112066894ab099b8c
<pre>
Return opaque origin for blob: URL containing inner non-http(s): URL
<a href="https://bugs.webkit.org/show_bug.cgi?id=257262">https://bugs.webkit.org/show_bug.cgi?id=257262</a>
rdar://109781193

Reviewed by Alex Christensen and Darin Adler.

Refactor SecurityOrigin so it is more clear blob: URLs are the sole
special case. And change how we derive the blob: URL origin to align
with the URL standard:

* No longer perform percent-decoding (matches other browsers).
* Restrict non-opaque origins to HTTP(S) URLs (will soon match other
  browsers). However:

  * Still give blob: URLs derived from file: origins an origin for now
    as removing that ability needs a bit more care. This currently goes
    against the URL standard, but that might change.
  * Also give registered schemes a pass to allow embedders to continue
    to use blob: URLs as they see fit.

Also change BlobURL to rely more directly on SecurityOrigin.

* LayoutTests/imported/w3c/web-platform-tests/url/a-element-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/a-element-origin-xhtml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any.worker-expected.txt:
* Source/WebCore/fileapi/BlobURL.cpp:
(WebCore::BlobURL::getOriginURL):
(WebCore::BlobURL::isSecureBlobURL):
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::getCachedOrigin):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::forBlobURL):
(WebCore::SecurityOrigin::isSecure):
(WebCore::SecurityOrigin::shouldUseInnerURL): Deleted.
(WebCore::SecurityOrigin::extractInnerURL): Deleted.
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::shouldTreatAsOpaqueOrigin):

Canonical link: <a href="https://commits.webkit.org/266247@main">https://commits.webkit.org/266247@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85470d2b0398da0cf676aeb8b3d62191b5e2e363

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15047 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12671 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16131 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13650 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15353 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14133 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11252 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15676 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11421 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12007 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19060 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12496 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12174 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15386 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12677 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10536 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11948 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3249 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16270 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12519 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/SecurityOriginData.cpp:
```diff
@@ -176,18 +176,17 @@ bool SecurityOriginData::shouldTreatAsOpaqueOrigin(const URL& url)
     // and could misinterpret another component for hostname.
-    if (schemeRequiresHost(innerURL) && innerURL.host().isEmpty())
+    if (schemeRequiresHost(extractedURL) && extractedURL.host().isEmpty())
         return true;
```
I think we can turn this into an assert with our new URL implementation. Happy to make that change if you agree or attempt as a follow-up.

achristensen07 commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -57,22 +57,9 @@ bool SecurityOrigin::shouldIgnoreHost(const URL& url)
 }
 
-bool SecurityOrigin::shouldUseInnerURL(const URL& url)
+URL SecurityOrigin::extractURLFromBlobURLPath(const URL& url)
```
This needs at least an ASSERT(url.protocolIsBlob()).
I think it would be safer to move the url.protocolIsBlob() to inside this function instead of requiring all callers to check before calling.

darinadler commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -57,22 +57,9 @@ bool SecurityOrigin::shouldIgnoreHost(const URL& url)
-{
-    // FIXME: Update this callsite to use the innerURL member function when
-    // we finish implementing it.
+    // FIXME: should this really be percent-decoding the path?
```
Can we construct test cases that help us answer this question?

annevk replied:
I added a test and other browsers don't have this behavior so I removed it, allowing for more simplification throughout the code.

annevk commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -179,18 +160,30 @@ SecurityOrigin::SecurityOrigin(const SecurityOrigin* other)
+
+    auto pathURL = URL({ }, url.path().toString());
+    // FIXME: blobinternal is web-exposed because of this
+    if (pathURL.isValid() && (pathURL.protocolIsInHTTPFamily() || pathURL.protocolIsFile() || pathURL.protocolIs("blobinternal"_s)))
```
"`blobinternal`" might not be needed. That's the last thought I had for some `FileReader`-related failures I cannot figure out.

sysrqb replied:
I'm not certain, but the WK2 debug bot may give a hint (https://ews-build.s3-us-west-2.amazonaws.com/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS/c74bcce5-6818/results.html):
```
inspector-resource:///TestCombined.js:2719:30: CONSOLE JS ERROR Cannot load blob:inspector-resource:///324f619c-6e99-425a-a1a3-fe3d74081459.
```

In the stderr logs.

annevk replied:
Thank you, that helped a lot!

annevk reviewed (commented):
This regresses `http/tests/inspector/network/fetch-response-body.html`, `http/tests/inspector/network/xhr-response-body.html`, and an API test, and I cannot figure out why. This morning I thought it might be about `blobinternal`, but that does not appear to be the case.

@cdumez @achristensen07 if you could provide me some assistance or are willing to take over I would greatly appreciate it.

darinadler commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -179,18 +160,36 @@ SecurityOrigin::SecurityOrigin(const SecurityOrigin* other)
+    if (auto origin = getCachedOrigin(url))
+        return origin.releaseNonNull();
+
+    auto pathURL = URL({ }, url.path().toString());
```
This is an old fashioned way to do it. The modern `URL` class does not require the `{ }`, and understands that a single string argument is an absolute URL. Should just be able to remove the `{ }`.

darinadler commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -179,18 +160,36 @@ SecurityOrigin::SecurityOrigin(const SecurityOrigin* other)
+        return origin.releaseNonNull();
+
+    auto pathURL = URL({ }, url.path().toString());
+    if (pathURL.isValid() && isSafelistedBlobProtocol(pathURL))
```
Maybe the `isValid` check can go inside `isSafelistedBlobProtocol`?

darinadler commented on Source/WebCore/page/SecurityOrigin.h:
```diff
@@ -47,7 +47,9 @@ class SecurityOrigin : public ThreadSafeRefCounted<SecurityOrigin> {
 
+    // https://url.spec.whatwg.org/#concept-url-origin
     WEBCORE_EXPORT static Ref<SecurityOrigin> create(const URL&);
+    WEBCORE_EXPORT static Ref<SecurityOrigin> forBlobURL(const URL&);
```
I think the name for this would be `createForBlobURL`. The name `forBlobURL` sounds like it will return the same one each time. We include `create` in names to make clear it’s new.

darinadler commented on Source/WebCore/page/SecurityOriginData.cpp:
```diff
@@ -176,18 +176,17 @@ bool SecurityOriginData::shouldTreatAsOpaqueOrigin(const URL& url)
-    // FIXME: Do we need to unwrap the URL further?
-    URL innerURL = SecurityOrigin::shouldUseInnerURL(url) ? SecurityOrigin::extractInnerURL(url) : url;
-    if (!innerURL.isValid())
+    auto urlOrBlobOriginURL = url.protocolIsBlob() ? BlobURL::getOriginURL(url) : url;
```
I prefer shorter names here. Could we come up with an adjective to use? I understand that `innerURL` is too abstract in a bad way. Maybe `policyURL` or `originURL` is a good name?

darinadler commented on Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
```diff
@@ -236,7 +237,7 @@ RefPtr<SecurityOrigin> ThreadableBlobRegistry::getCachedOrigin(const URL& url)
         return cachedOrigin;
 
-    if (!url.protocolIsBlob() || !isBlobURLContainsNullOrigin(url))
+    if (!isBlobURLContainsNullOrigin(url))
```
Maybe at some point we should rename to `isBlobURLContainingNullOrigin`.

annevk replied:
https://bugs.webkit.org/show_bug.cgi?id=259426

achristensen07 commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -179,18 +160,36 @@ SecurityOrigin::SecurityOrigin(const SecurityOrigin* other)
+{
+    return url.protocolIsInHTTPFamily()
+        || url.protocolIsFile()
+        || LegacySchemeRegistry::schemeIsHandledBySchemeHandler(url.protocol());
```
This is process-global.  Could you add an assertion that it is not used in the network process or gpu process?  Otherwise we'll have scope issues like the one I fixed in 263652@main

annevk replied:
It would really help me if you could explain this more. As well as why this is not a problem for `SecurityOriginData::shouldTreatAsOpaqueOrigin`.

The asserts cause everything to fail so clearly something else is needed here.

darinadler commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -179,18 +160,48 @@ SecurityOrigin::SecurityOrigin(const SecurityOrigin* other)
-        return cachedOrigin.releaseNonNull();
+    // FIXME: We ought to be able to assert that isInWebProcess() is true, but
+    // WebKitLegacy doesn't have a web content process.
+    // This is for LegacySchemeRegistry::schemeIsHandledBySchemeHandler.
```
Something like this seems guaranteed to get out of sync. I wish there was a way to out the assertions right at the code that uses the global state but also have the benefit of checking when we take other paths. I guess we can leave like this for now.

annevk replied:
Unfortunately this doesn't seem to work at all. See all the failures on the debug bot. Not sure how to proceed with this.

annevk committed:
Return opaque origin for blob: URL containing inner non-http(s): URL
https://bugs.webkit.org/show_bug.cgi?id=257262
rdar://109781193

Reviewed by Alex Christensen and Darin Adler.

Refactor SecurityOrigin so it is more clear blob: URLs are the sole
special case. And change how we derive the blob: URL origin to align
with the URL standard:

* No longer perform percent-decoding (matches other browsers).
* Restrict non-opaque origins to HTTP(S) URLs (will soon match other
  browsers). However:

  * Still give blob: URLs derived from file: origins an origin for now
    as removing that ability needs a bit more care. This currently goes
    against the URL standard, but that might change.
  * Also give registered schemes a pass to allow embedders to continue
    to use blob: URLs as they see fit.

Also change BlobURL to rely more directly on SecurityOrigin.

* LayoutTests/imported/w3c/web-platform-tests/url/a-element-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/a-element-origin-xhtml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/url/url-origin.any.worker-expected.txt:
* Source/WebCore/fileapi/BlobURL.cpp:
(WebCore::BlobURL::getOriginURL):
(WebCore::BlobURL::isSecureBlobURL):
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::getCachedOrigin):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::forBlobURL):
(WebCore::SecurityOrigin::isSecure):
(WebCore::SecurityOrigin::shouldUseInnerURL): Deleted.
(WebCore::SecurityOrigin::extractInnerURL): Deleted.
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::shouldTreatAsOpaqueOrigin):

Canonical link: https://commits.webkit.org/266247@main
